pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -374,20 +381,60 @@ func (m *kubeGenericRuntimeManager) podSandboxChanged(pod *api.Pod, podStatus *k
 	return false, sandboxStatus.Metadata.GetAttempt(), sandboxStatus.GetId()
 }
 
+// checkAndKeepInitContainers keeps all successfully completed init containers. If there
+// are failing containers, only keep the first failing one.
+func checkAndKeepInitContainers(pod *api.Pod, podStatus *kubecontainer.PodStatus, initContainersToKeep map[kubecontainer.ContainerID]int) bool {
+	initFailed := false
+
+	for i, container := range pod.Spec.InitContainers {
+		containerStatus := podStatus.FindContainerStatusByName(container.Name)
+		if containerStatus == nil {
+			continue
+		}
+
+		if containerStatus.State == kubecontainer.ContainerStateRunning {
+			initContainersToKeep[containerStatus.ID] = i
+			continue
+		}
+
+		if containerStatus.State == kubecontainer.ContainerStateExited {
+			initContainersToKeep[containerStatus.ID] = i
+			if containerStatus.ExitCode != 0 {
+				initFailed = true
+				break
+			}
+		}
+	}
+
+	return initFailed
+}
+
 // computePodContainerChanges checks whether the pod spec has changed and returns the changes if true.
 func (m *kubeGenericRuntimeManager) computePodContainerChanges(pod *api.Pod, podStatus *kubecontainer.PodStatus) podContainerSpecChanges {
 	glog.V(5).Infof("Syncing Pod %q: %+v", format.Pod(pod), pod)
 
 	sandboxChanged, attempt, sandboxID := m.podSandboxChanged(pod, podStatus)
 	changes := podContainerSpecChanges{
-		CreateSandbox:     sandboxChanged,
-		SandboxID:         sandboxID,
-		Attempt:           attempt,
-		ContainersToStart: make(map[int]string),
-		ContainersToKeep:  make(map[kubecontainer.ContainerID]int),
-		ContainersToKill:  make(map[kubecontainer.ContainerID]containerToKillInfo),
-	}
-
+		CreateSandbox:        sandboxChanged,
+		SandboxID:            sandboxID,
+		Attempt:              attempt,
+		ContainersToStart:    make(map[int]string),
+		ContainersToKeep:     make(map[kubecontainer.ContainerID]int),
+		InitContainersToKeep: make(map[kubecontainer.ContainerID]int),
+		ContainersToKill:     make(map[kubecontainer.ContainerID]containerToKillInfo),
+	}
+
+	// check the status of init containers.
+	initFailed := false
+	// always reset the init containers if the sandbox is changed.
+	if !sandboxChanged {
+		// Keep all successfully completed containers. If there are failing containers,
+		// only keep the first failing one.
+		initFailed = checkAndKeepInitContainers(pod, podStatus, changes.InitContainersToKeep)
+	}
+	changes.InitFailed = initFailed
+
+	// check the status of containers.
```

Comment:
If init containers failed, why should we even bother check the Spec.Containers???